drivers: add IS31FL3194 LED driver + use it for Arduino Nicla Sense ME

pillo79 opened the pull request:
This PR introduces support for the RGB LED on the Arduino Nicla Sense ME board in 3 steps:
* a generic driver for the ISSI IS31FL3194 3-channel LED driver;
* devicetree and config changes to enable it on the Arduino Nicla Sense ME;
* a sample application for hardware that has that.

pillo79 committed:
drivers/led: add IS31FL3194 LED driver

Add support for the IS31FL3194 3-channel LED driver. This driver can be
configured to handle its outputs as either a single RGB LED or (up to)
three independent LEDs.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
samples: add IS31FL3194 RGB LED driver sample

This sample app cycles several colors on an RGB LED forever using the LED
API. It can be built and executed on an Arduino Nicla Sense ME, or on any
board where the devicetree has a similar compatible I2C device node.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
arduino_nicla_sense_me: add RGB LED support to the device tree

This patch enables the user to control the RGB LED on the Arduino Nicla
Sense ME board. The RGB LED is controlled by the IS31FL3194 LED driver
and is connected to the I2C0 bus.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

decsny commented on dts/bindings/led/issi,is31fl3194.yaml:
```diff
@@ -0,0 +1,78 @@
+        - 40
+      default: 20
+      description: |
+        The current limit for the LED in mA. The default value is 20mA.
```
default values in DT bindings need to have justifications in the description

pillo79 commented:
v2:
* Reworded description to explain the default LED current used is the hardware default.

simonguinot commented:
Here is a previous attempt to merge a driver for the IS31FL3194 controller: #45915

decsny commented on dts/bindings/led/issi,is31fl3194.yaml:
```diff
@@ -0,0 +1,79 @@
+      default: 20
+      description: |
+        The current limit for the LED in mA. If not specified,
+        the hardware default value of 20mA is used.
```
a justification is a "why"

https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

simonguinot reviewed (changes requested):
Hi @pillo79,

Thanks for contributing this driver. It looks mostly good to me. Please find my review below.

pillo79 commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,305 @@
+	if (led_info->num_colors != 3) {
+		return -ENOTSUP;
+	}
+	if (num_colors != 3) {
```
Do I? This API does not provide the "start offset" so you are not free to edit a single specific channel; I could only allow editing subsets from index 0 onwards. I feel this would be an "unintended feature" of the API :slightly_smiling_face: 

This also raises a related question: what is the data ordering here? How should color mapping be handled? AFAICS:
* "led" drivers _ignore_ color mapping info and leave it to the user to provide input in the correct order; 
* "led_strip" drivers map RGB-ordered data (from `struct led_rgb`) to the appropriate channels using the color mapping info.

I did implement the latter without much thinking about it, since I assumed that was the intention of the `color-mapping` info in the DTS. Why should the user have to deal with this in their app?

simonguinot replied:
> Do I? This API does not provide the "start offset" so you are not free to edit a single specific channel; I could only allow editing subsets from index 0 onwards. I feel this would be an "unintended feature" of the API 🙂

Yes you are right. It is actually the purpose of the `->write_channels()` API (not implemented in your driver) to allow specific channels to be updated.

> 
> This also raises a related question: what is the data ordering here? How should color mapping be handled? AFAICS:
> 
>     * "led" drivers _ignore_ color mapping info and leave it to the user to provide input in the correct order;

Yes it is assumed that all the colors of a single LED are passed to `->set_color()` following the order defined by the `color-mapping` DT property.

> 
>     * "led_strip" drivers map RGB-ordered data (from `struct led_rgb`) to the appropriate channels using the color mapping info.

Yes. And it is something I'd like to see unified with the idea of merging the led and led_strip subsystems.

> 
> I did implement the latter without much thinking about it, since I assumed that was the intention of the `color-mapping` info in the DTS. Why should the user have to deal with this in their app?

And that's correct :)

pillo79 commented:
**v2**:
* Grouped config checks to the `is31fl3194_check_config` function, including: 
  * Binding-time check for `index` values (only 0, 1, or 2 allowed).
  * Build-time test for at least one LED definition per instance.
  * Detect multiple LEDs with same output index.
  * Detect invalid RGB LED color definitions.
  * Correctly handle 1 single LED configuration.
* Made `current_limit` required, as it's a board-dependent hardware parameter.

simonguinot commented:
Hi @pillo79. Thanks for the update. I am off this week, so I'll review it next week.

simonguinot commented:
@bbilas @kartben please can you have a look ?

simonguinot commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,305 @@
+		for (i = 0; i < IS31FL3194_CHANNEL_COUNT; i++) {
+			current_reg |= band << (2 * i);
+		}
+	} else if (config->num_leds <= IS31FL3194_CHANNEL_COUNT) {
```
Thanks. It is much better now.

simonguinot reviewed (commented):
Hi @pillo79.

It is looking good to me. I'll do a final pass once others have added their review.

bbilas commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,350 @@
+			break;
+		}
+	}
+	if (!ret) {
```
ret is not a bool,  see: https://pvs-studio.com/en/docs/warnings/v2584/ and fix everything according to that.

bbilas commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,350 @@
+
+	if (config->num_leds > IS31FL3194_CHANNEL_COUNT) {
+		LOG_ERR("%s: too many LEDs defined", dev->name);
+		return -EINVAL;
```
I would use `ENOTSUP` here.

bbilas commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,350 @@
+		if (info->index != 0) {
+			LOG_ERR("%s: channel index must be 0 with an RGB LED", dev->name);
+			return -EINVAL;
+		}
```
Please add nl after each "if" condition.

bbilas commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,350 @@
+
+	info = &config->led_infos[0];
+	if ((config->num_leds == 1) && (info->num_colors == 3)) {
+
```
Useless nl.

bbilas commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,350 @@
+		}
+
+	} else {
+
```
ditto

bbilas commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,350 @@
+				return -EINVAL;
+			}
+		}
+
```
ditto

pillo79 commented:
Thanks everyone for your review.

v3:
- Fully spelled out bool-type checks
- Removed unused register/mask `#define`s
- Improved DTS comments
- (cosmetic) Fixed the sample project name
- (cosmetic) Applied proposed whitespace fixes

simonguinot commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,350 @@
+				return -EINVAL;
+			}
+
+			used_idxs[info->index]++;
```
And here you should rather check that the three color channels are all used.

simonguinot reviewed (changes requested):
Hi @pillo79,

Thanks for the update. It is much better now and we are almost there. Please find my final review below.

pillo79 commented on dts/bindings/led/issi,is31fl3194.yaml:
```diff
@@ -0,0 +1,85 @@
+      required: true
+
+    index:
+      required: true
```
Yes, I was thinking about supporting a board with, say, only one LED connected to the second channel of the IC. But it's not really a plausible scenario, if you design with this component you need all 3 channels. Will fix.

pillo79 commented:
v3:
- rewrote driver checks so a full RGB connection is enforced (either via 3 RGB LEDs or 1 3-channel LED)
- removed every use of the `index` DT property 
- applied other review comments

simonguinot commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,357 @@
+	switch (color_id) {
+	case LED_COLOR_ID_RED:
+		rgb_counts[0]++;
+		return true;
```
At the expense of the `multiple channels map to ...` error message you can:

`return (++rgb_counts[0] == 1);`

Choose the option you prefer.

simonguinot commented on drivers/led/is31fl3194.c:
```diff
@@ -0,0 +1,357 @@
+	const struct is31fl3194_config *config = dev->config;
+	const struct led_info *info;
+	int rgb_counts[3] = { 0 };
+	int i;
```
you can use `uint8_t` for the two variables above.

simonguinot reviewed (commented):
It looks good to me. Please find my last comments below.

simonguinot commented:
Hi @bbilas. Please can you revisit ?

pillo79 commented:
v5: 
* applied review comments

simonguinot reviewed (approved):
Thanks for this nice driver @pillo79 !
